### ISSUE TITLE
How does the ECU decide when to use MAP estimate?

### ISSUE THREAD

**Issue by ElDominio** [rusefi/rusefi#7030]
**Describe the bug**
It is not clear when the ECU uses the MAP estimate table. Are there thresholds to use it? Are they user-configurable?

**To Reproduce**
1. Have a MAP that reads 0.1v
2. MAP shown is 0kpa
3. Should be using MAP estimate from estimate table

**Expected behavior**
When MAP is considered failed (who decides this?), we should fallback to MAP estimate when running Speed Density.

**Comment by FDSoftware** [comment]
**TLDR**: rusEFI (for now) does not decide in any (practical) case the use of the MAP estimate table

nerd research below:

checking:
`firmware/controllers/algo/airmass/speed_density_airmass.cpp`
```c++
float SpeedDensityAirmass::getMap(float rpm, bool postState) const {
	float fallbackMap = m_mapEstimationTable->getValue(rpm, Sensor::getOrZero(SensorType::Tps1));

#if EFI_TUNER_STUDIO
	if (postState) {
		engine->outputChannels.fallbackMap = fallbackMap;
	}
#endif // EFI_TUNER_STUDIO

	return Sensor::get(SensorType::Map).value_or(fallbackMap);
}

``` 
and:
`firmware/controllers/engine_cycle/map_averaging.cpp`
```c++
#if HAL_USE_ADC

/**
 * This method is invoked from ADC callback.
 * @note This method is invoked OFTEN, this method is a potential bottleneck - the implementation should be
 * as fast as possible
 */
void mapAveragingAdcCallback(float instantVoltage) {
	efiAssertVoid(ObdCode::CUSTOM_ERR_6650, hasLotsOfRemainingStack(), "lowstck#9a");

	SensorResult mapResult = getMapAvg(currentMapAverager).submit(instantVoltage);

	if (!mapResult) {
		// hopefully this warning is not too much CPU consumption for fast ADC callback
		warning(ObdCode::CUSTOM_INSTANT_MAP_DECODING, "Invalid MAP at %f", instantVoltage);
	}

#if EFI_TUNER_STUDIO
	float instantMap = mapResult.value_or(0);
	engine->outputChannels.instantMAPValue = instantMap;
#endif // EFI_TUNER_STUDIO
}
#endif
``` 
and:
`firmware/init/sensor/init_map.cpp`
```c++
// lowest reasonable idle is maybe 600 rpm
// one sample per cycle (1 cylinder, or "sample one cyl" mode) gives a period of 100ms
// add some margin -> 200ms timeout for fast MAP sampling
MapAverager fastMapSensor(SensorType::MapFast, MS2NT(200));
MapAverager fastMapSensor2(SensorType::MapFast2, MS2NT(200));

MapAverager& getMapAvg(size_t idx) {
	return idx == 0 ? fastMapSensor : fastMapSensor2;
}

``` 
and (finally):
`firmware/controllers/engine_cycle/map_averaging.cpp`

```c++
SensorResult MapAverager::submit(float volts) {
	auto result = m_function ? m_function->convert(volts) : unexpected;

	if (m_isAveraging && result) {
		chibios_rt::CriticalSectionLocker csl;

		m_counter++;
		m_sum += result.Value;
	}

	return result;
}

``` 
The estimated map will be used in cases where the sensor is not initialized, since the low/high voltages are not checked throughout the implementation, although they are defined in other parts of the code. I think the value of 0Kpa in TunerStudio is due to this section:

```c++
#if EFI_TUNER_STUDIO
	float instantMap = mapResult.value_or(0);
	engine->outputChannels.instantMAPValue = instantMap;
#endif // EFI_TUNER_STUDIO
```

**Comment by rusefillc** [comment]
Invalid MAP detection could be related to the below

![image](https://github.com/user-attachments/assets/8a189d10-97cd-4be4-95c9-31e6c8e6d4c6)

**Comment by ElDominio** [comment]
> Invalid MAP detection could be related to the below
> 
> ![image](https://github.com/user-attachments/assets/8a189d10-97cd-4be4-95c9-31e6c8e6d4c6)
> 

It used to be? I know I have used that successfully but I could not get it to work today

**Comment by rusefillc** [comment]
@ElDominio just added new logging data point / trying to add explicit indicator right on that dialog https://github.com/rusefi/rusefi/commit/ed8fbde6e7b722579e951e8c3186db595c45f76f

**Comment by rusefillc** [comment]
@ElDominio colors are odd but text seems to change based on thresholds

![image](https://github.com/user-attachments/assets/083379a6-ef33-41b2-aedf-102732c50902)

![image](https://github.com/user-attachments/assets/8f220d24-0618-48c2-8c88-1888966c99a6)

**Comment by ElDominio** [comment]
Change "happy" to "good" for professional reasons
